README.md
# Flexitime

Flexitime is a Ruby date/time string parser with the intended purpose of converting a string value received from a UI or API into an [ActiveSupport::TimeWithZone](https://api.rubyonrails.org/classes/ActiveSupport/TimeWithZone.html) object. It offers the flexibility of deciphering date/time strings in the most common formats with the ability to self-determine the expected order of the date parts when using [rails-i18n](https://github.com/svenfuchs/rails-i18n). The date order and a desired precision for the time object can also be set via configuration options or method arguments.

The gem was born of the need to parse date, datetime & time strings in a multi-user environment supporting different locales and time zones. Depending upon the user's locale the UI would return date/time strings in different formats and in different orders (day/month/year or month/day/year). This variation in the ordering of the day and month parts proved to be the main catalyst to finding or creating a date/time parser. The resultant time object needed to be created in the user's time zone and additionally the system stored times only to a minute precision. Flexitime was created to provide a simple yet flexible parser to meet these needs.

![Build Status](https://github.com/CircleSD/flexitime/actions/workflows/ci.yml/badge.svg?branch=main)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'flexitime'
```

And then execute:

```ruby
bundle install
```

Or install it yourself as:

```ruby
gem install flexitime
```

## Usage

The Flexitime `parse` method accepts a string argument or an object that implements the `to_str` method (to denote that it behaves like a string). When the string is recognised as a valid date, datetime or time it returns an [ActiveSupport::TimeWithZone](https://api.rubyonrails.org/classes/ActiveSupport/TimeWithZone.html) object. If any of the date/time parts are invalid or the string does not match a recognised format the method returns `nil`.

As Flexitime uses the [TimeZone](https://api.rubyonrails.org/classes/ActiveSupport/TimeZone.html) class, `Time.zone` should be set before using the `parse` method. If `Time.zone` is `nil` Flexitime will set the zone to UTC.

```ruby
Time.zone = "London"
Flexitime.parse("23/08/2021")       # => Mon, 23 Aug 2021 00:00:00.000000000 BST +01:00
Flexitime.parse("23/08/2021 08:00") # => Mon, 23 Aug 2021 08:00:00.000000000 BST +01:00
Flexitime.parse("08:00")            # => Mon, 14 Mar 2022 08:00:00.000000000 GMT +00:00
```

```ruby
Time.zone = "London"
Flexitime.parse("31/02/2021 08:00") # => nil
Flexitime.parse("computer says no") # => nil
```

### First Date Part

The `first_date_part` option is used to denote the first part of the date within the string, either `:day` or `:month`. When using the [rails-I18n gem](https://github.com/svenfuchs/rails-i18n) if the first element of the "date.order" translation is :day or :month that will be used. Alternatively the option can be set manually via a configuration option or a `parse` argument. If the option has not been set then the default of `:day` is used.

```ruby
Time.zone = "London"
I18n.locale = :"en-GB"
I18n.t("date.order")          # => [:day, :month, :year]
Flexitime.parse("01/06/2021") # => Tue, 01 Jun 2021 00:00:00.000000000 BST +01:00

I18n.locale = :"en-US"
I18n.t("date.order")          # => [:month, :day, :year]
Flexitime.parse("01/06/2021") # => Wed, 06 Jan 2021 00:00:00.000000000 GMT +00:00
```

```ruby
Time.zone = "London"
Flexitime.first_date_part = :day
Flexitime.parse("01/06/2021") # => Tue, 01 Jun 2021 00:00:00.000000000 BST +01:00

Flexitime.first_date_part = :month
Flexitime.parse("01/06/2021") # => Wed, 06 Jan 2021 00:00:00.000000000 GMT +00:00
```

```ruby
Time.zone = "London"
Flexitime.parse("01/06/2021", first_date_part: :day)   # => Tue, 01 Jun 2021 00:00:00.000000000 BST +01:00
Flexitime.parse("01/06/2021", first_date_part: :month) # => Wed, 06 Jan 2021 00:00:00.000000000 GMT +00:00
```

Regardless of the `first_date_part` value, Flexitime will always attempt to parse a string starting with a 4 digit year, deciphering the format as year/month/day.

```ruby
Time.zone = "London"
Flexitime.parse("2021-11-12 08:15") # => Fri, 12 Nov 2021 08:15:00.000000000 GMT +00:00
```

### Precision

The `precision` option denotes the desired precision for the returned time objects. This defaults to minute (`:min`) meaning that the time object will be returned without seconds. The option can be set via a configuration option or a `parse` argument and the accepted values are `:day`, `:hour`, `:min`, `:sec` or `:usec`

```ruby
Time.zone = "London"
Flexitime.precision = :day
Flexitime.parse("2022-12-01T18:30:45.036711Z")  # => Thu, 01 Dec 2022 00:00:00.000000000 GMT +00:00
Flexitime.precision = :hour
Flexitime.parse("2022-12-01T18:30:45.036711Z")  # => Thu, 01 Dec 2022 18:00:00.000000000 GMT +00:00
Flexitime.precision = :min
Flexitime.parse("2022-12-01T18:30:45.036711Z")  # => Thu, 01 Dec 2022 18:30:00.000000000 GMT +00:00
Flexitime.precision = :sec
Flexitime.parse("2022-12-01T18:30:45.036711Z")  # => Thu, 01 Dec 2022 18:30:45.000000000 GMT +00:00
Flexitime.precision = :usec
Flexitime.parse("2022-12-01T18:30:45.036711Z")  # => Thu, 01 Dec 2022 18:30:45.036711000 GMT +00:00

Flexitime.parse("2022-12-01T18:30:45.036711Z", precision: :day)  # => Thu, 01 Dec 2022 00:00:00.000000000 GMT +00:00
Flexitime.parse("2022-12-01T18:30:45.036711Z", precision: :hour) # => Thu, 01 Dec 2022 18:00:00.000000000 GMT +00:00
Flexitime.parse("2022-12-01T18:30:45.036711Z", precision: :min)  # => Thu, 01 Dec 2022 18:30:00.000000000 GMT +00:00
Flexitime.parse("2022-12-01T18:30:45.036711Z", precision: :sec)  # => Thu, 01 Dec 2022 18:30:45.000000000 GMT +00:00
Flexitime.parse("2022-12-01T18:30:45.036711Z", precision: :usec) # => Thu, 01 Dec 2022 18:30:45.036711000 GMT +00:00
```

### Ambiguous Year Future Bias

The `ambiguous_year_future_bias` configuration option is used to determine the century when parsing a string containing a 2 digit year and defaults to 50.

With a bias of 50, when the current year is 2020 the time object's year is set from within a range of >= 1970 and <= 2069

With a bias of 20, when the current year is 2020 the time object's year is set from within a range of >= 2000 and <= 2099

```ruby
Time.zone = "London"
Flexitime.parse("01/08/00").year # => 2000
Flexitime.parse("01/08/71").year # => 2071
Flexitime.parse("01/08/72").year # => 1972
Flexitime.parse("01/08/99").year # => 1999
```

### Thread Safety

The Flexitime `configuration` instance is stored in the currently executing thread as a thread-local variable in order to avoid conflicts with concurrent requests and make the variable threadsafe.

## Formats

Flexitime uses regular expressions to match the string to the most common date & time formats and uses the matched parts to create a time object.

Dates separated by either forward slashes `/` hyphens `-` or periods `.`; with 1 or 2 digit days and months; with 2 or 4 digit years; and with day/month or year first (always a 4 digit year).

```text
01/08/2021
01-08-2021
01.08.2021
1/8/2021
1-8-2021
1.8.2021
01/08/21
01-08-21
01.08.21
1/8/21
1-8-21
1.8.21
2021/08/01
2021-08-01
2021.08.01
2021/8/1
2021-8-1
2021.8.1
```

Times with hours, minutes & seconds separated by colons `:`; with hours & minutes separated by either colons `:` or periods `.`; with 1 or 2 digit hours; and with case-insensitive AM or PM.

```text
08:15:30
08:15
08.15
8:15:30
8.15
08:15:30 PM
08:15 am
```

Also times in the ISO 8601 Zulu format with between 1 and 6 digit milliseconds

```text
2021-08-01T08:15:30.144515Z
```

If the string does not match any of the regular expressions then Flexitime will attempt to parse the string using the TimeZone class, so you lose nothing from using Flexitime and it will still return a time object for a string containing, for example, an offset or words.

```ruby
Time.zone = "Kyiv"
Flexitime.parse("2022-02-24T09:00:00+02:00")  # => Thu, 24 Feb 2022 09:00:00.000000000 EET +02:00
Flexitime.parse("2nd January 2021")           # => Sat, 02 Jan 2021 00:00:00.000000000 EET +02:00
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

After making changes run `rake spec` to run the tests and `bundle exec appraisal rspec` to run the tests against different versions of activesupport; and run `bundle exec standardrb` to check the style of files.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on [GitHub](https://github.com/CircleSD/flexitime). This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/CircleSD/flexitime/blob/main/CODE_OF_CONDUCT.md).

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## Credits

Tom Preston-Werner and [Chronic](https://github.com/mojombo/chronic) which was our previous go-to parser of choice but is unfortunately no longer maintained and performs natural language parsing that we do not require. It was a useful reference in particular for the `ambiguous_year_future_bias` configuration options.

Adam Meehan and [Timeliness](https://github.com/adzap/timeliness) which was a close match for our needs but proved to be too strict in its accepted formats particulary as we wanted to cater for a variety of date separators for both day/month/year and month/day/year date ordering. The gem provided some very useful inspiration in regards to code structure with forwarding for the configuration class and testing thread safety.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Flexitime project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/CircleSD/flexitime/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
# Change Log

## 1.0.0 - 2022-03-17

* Perform parsing using `Time.zone` ([d30b224](https://github.com/CircleSD/flexitime/commit/d30b224f40379331473f60d11f6dcc8b1875413d))
* Add `first_date_part` and `precision` arguments to `parse` method ([18ce552](https://github.com/CircleSD/flexitime/commit/18ce5528caf2f1c6d2643b610a713e7d2cd6ce75))

## 0.1.0 - 2021-11-22

* Initial release


spec/spec_helper.rb
# frozen_string_literal: true

# Initiate code coverage but not when running appraisal
# to prevent an error with activesupport 4.0
# View test coverage report after running test suite by using the following command
# > open coverage/index.html
# https://github.com/simplecov-ruby/simplecov
if ENV["APPRAISAL_INITIALIZED"].nil?
  require "simplecov"
  SimpleCov.start do
    add_filter "/spec/"
  end
end

require "flexitime"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end

I18n.available_locales = ["en", "en-GB", "en-US", "en-IE", "nl"]


benchmark.rb
# Benchmark file to compare Flexitime with Time and Time.zone parsing
# to ensure Flexitime performs as well as possible
# To perform benchmark comparison run:-
# $ ruby benchmark.rb
$:.unshift(File.expand_path("lib"))

require "benchmark"
require "flexitime"
require "active_support/time"

Time.zone = "Europe/London"

n = 10_000

Benchmark.bm(50) do |benchmark|
  benchmark.report("Time.parse datetime YMD") do
    n.times do
      Time.parse("2021-08-12 12:30")
    end
  end

  benchmark.report("Time.zone.parse datetime YMD") do
    n.times do
      Time.zone.parse("2021-08-12 12:30")
    end
  end

  benchmark.report("Flexitime.parse datetime YMD") do
    n.times do
      Flexitime.parse("2021-08-12 12:30")
    end
  end
end

Benchmark.bm(50) do |benchmark|
  benchmark.report("Time.parse datetime DMY HM") do
    n.times do
      Time.parse("23/08/2021 12:30")
    end
  end

  benchmark.report("Time.zone.parse datetime DMY HM") do
    n.times do
      Time.zone.parse("23/08/2021 12:30")
    end
  end

  benchmark.report("Flexitime.parse datetime DMY HM") do
    n.times do
      Flexitime.parse("23/08/2021 12:30")
    end
  end
end

Benchmark.bm(50) do |benchmark|
  benchmark.report("Time.parse datetime DMY HMS") do
    n.times do
      Time.parse("23/08/2021 12:30:45")
    end
  end

  benchmark.report("Time.zone.parse datetime DMY HMS") do
    n.times do
      Time.zone.parse("23/08/2021 12:30:45")
    end
  end

  benchmark.report("Flexitime.parse datetime DMY HMS") do
    n.times do
      Flexitime.parse("23/08/2021 12:30:45")
    end
  end
end

Benchmark.bm(50) do |benchmark|
  benchmark.report("Time.parse time HM") do
    n.times do
      Time.parse("12:30")
    end
  end

  benchmark.report("Time.zone.parse time HM") do
    n.times do
      Time.zone.parse("12:30")
    end
  end

  benchmark.report("Flexitime.parse time HM") do
    n.times do
      Flexitime.parse("12:30")
    end
  end
end

Benchmark.bm(50) do |benchmark|
  benchmark.report("Time.parse time HMS") do
    n.times do
      Time.parse("12:30:45")
    end
  end

  benchmark.report("Time.zone.parse time HMS") do
    n.times do
      Time.zone.parse("12:30:45")
    end
  end

  benchmark.report("Flexitime.parse time HMS") do
    n.times do
      Flexitime.parse("12:30:45")
    end
  end
end


lib/flexitime.rb
# frozen_string_literal: true

require "time"
require "forwardable"

# Active Support Core Extensions
# https://guides.rubyonrails.org/active_support_core_extensions.html
begin
  # workaround for activesupport 7.0
  # ref: https://github.com/rails/rails/issues/43851
  require "active_support/isolated_execution_state"
rescue LoadError
end
require "active_support/time" # Time.zone
require "active_support/core_ext/object/blank" # blank? and present?
require "active_support/core_ext/array/access" # array second/third/fourth
require "active_support/core_ext/time/calculations" # Time.days_in_month

require_relative "flexitime/version"
require_relative "flexitime/configuration"
require_relative "flexitime/parser"

module Flexitime
  class << self
    extend Forwardable
    def_delegators :configuration, :first_date_part, :precision, :ambiguous_year_future_bias
    def_delegators :configuration, :first_date_part=, :precision=, :ambiguous_year_future_bias=

    include Flexitime::Parser

    def configuration
      Thread.current["Flexitime.configuration"] ||= Configuration.new
    end
  end
end


lib/flexitime/parser.rb
# frozen_string_literal: true

module Flexitime
  module Parser
    # date regex to match year/month/day eg. "2021-10-27"
    # with 4 digit year; 1 or 2 digit month & day; and slash, hyphen or period separator
    ISO8601_DATE_REGEX = %r{(?<!\d)(\d{4})[-./](\d{1,2})[-./](\d{1,2})(?!\d)}
    # date regex to match day/month/year and month/day/year eg. "27/10/2021" and "10/27/2021"
    # with 1 or 2 digit day & month; 2 or 4 digit year; and slash, hyphen or period separator
    LOCAL_DATE_REGEX = %r{(?<!\d)(\d{1,2})[-./](\d{1,2})[-./](\d{4}|\d{2})(?!\d)}
    # time regex to match a zulu time "11:14:30.000Z"
    # with 2 digit hour, minute & second; 1 to 6 digit milliseconds; and Z character
    ISO8601_TIME_REGEX = %r{(\d{2}):(\d{2}):(\d{2})\.(\d{1,6})(Z{1})$}
    # time regex to match hour/min/sec eg. "09:15:30", "09:15:30 AM" and "09:15:30AM"
    # with 1 or 2 digit hour; 2 digit minute & second; optional case-insensitive meridiem;
    # and colon separator but not a period separator as there are no rails-i18n locales
    # using that separator with hours, minutes & seconds and it would clash with the local date pattern
    HOUR_MINUTE_SECOND_REGEX = %r{^(\d{1,2}):(\d{2}):(\d{2})\s?([aApP][mM])?\.?$}
    # time regex to match hour/min eg. "09:15", "09:15 AM" and "09:15AM"
    # with 1 or 2 digit hour; 2 digit minute; optional case-insensitive meridiem;
    # and colon or period separator as in rails-i18n there a few locales (such as Danish)
    # that use a period with hours & minutes
    HOUR_MINUTE_REGEX = %r{^(\d{1,2})[:.](\d{2})\s?([aApP][mM])?\.?$}

    # Parse a String argument and create an ActiveSupport::TimeWithZone object using Time.zone
    #
    # The parse uses either the argument or configuration first_date_part (:day or :month)
    # and either the argument or configuration precision (:day, :hour, :min, :sec or :usec).
    #
    # When the String contains a date or date/time or time that matches the regular expressions
    # the Time.zone local method is used to create a TimeWithZone object
    # otherwise the Time.zone parse method is used to create a TimeWithZone object
    # and for an invalid date, date/time or time nil is returned.
    def parse(str, first_date_part: nil, precision: nil)
      validate_options(first_date_part: first_date_part, precision: precision)

      str = str.is_a?(String) ? str : str.try(:to_str)
      return nil if str.blank?

      parts = extract_parts(str, first_date_part: first_date_part)

      if parts.present?
        create_time_from_parts(parts, precision: precision) if valid_date_parts?(parts)
      else
        time_zone_parse(str, precision: precision)
      end
    end

    private

    # Returns an ActiveSupport::TimeZone
    def time_zone
      Time.zone ||= "UTC"
      Time.zone
    end

    # Ensure the parse method options are valid
    # reusing the configuration class validation which raises an exception
    def validate_options(first_date_part: nil, precision: nil)
      return if first_date_part.blank? && precision.blank?
      config = Configuration.new
      config.first_date_part = first_date_part if first_date_part.present?
      config.precision = precision if precision.present?
    end

    # Extract date and time parts and return a Hash containing the parts
    # or nil if either the date or time string does not match a regex
    def extract_parts(str, first_date_part: nil)
      date_str, time_str = separate_date_and_time(str)

      date_parts = extract_date_parts(date_str, first_date_part: first_date_part)

      if date_parts.blank?
        now = time_zone.now
        date_parts = {year: now.year, month: now.month, day: now.day}
        time_str = str.strip
      end

      if time_str.present?
        time_parts = extract_time_parts(time_str)
        return nil if time_parts.blank?
        date_parts.merge(time_parts)
      else
        date_parts
      end
    end

    def separate_date_and_time(str)
      parts = str.index(" ").present? ? str.split(" ") : str.split("T")
      [parts.shift, parts.join(" ")]
    end

    def extract_date_parts(str, first_date_part: nil)
      extract_iso_date_parts(str) || extract_local_date_parts(str, first_date_part: first_date_part)
    end

    def extract_iso_date_parts(str)
      # match array returns ["1973-08-23", "1973", "08", "23"]
      parts = str.match(ISO8601_DATE_REGEX).to_a.pop(3).map(&:to_i)
      if parts.present?
        {year: make_year(parts.first), month: parts.second, day: parts.third}
      end
    end

    def extract_local_date_parts(str, first_date_part: nil)
      # match array returns ["23-08-1973", "23", "08", "1973"]
      parts = str.match(LOCAL_DATE_REGEX).to_a.pop(3).map(&:to_i)
      if parts.present?
        first_date_part ||= Flexitime.configuration.first_date_part
        day, month = first_date_part == :day ? [parts.first, parts.second] : [parts.second, parts.first]
        {year: make_year(parts.third), month: month, day: day}
      end
    end

    def extract_time_parts(str)
      extract_iso_time_parts(str) || extract_hour_minute_second_parts(str) || extract_hour_minute_parts(str)
    end

    def extract_iso_time_parts(str)
      # match array returns ["11:14:30.999Z", "11", "14", "30", "999", "Z"]
      parts = str.match(ISO8601_TIME_REGEX).to_a.pop(5).map(&:to_i)
      if parts.present?
        {hour: parts.first, min: parts.second, sec: parts.third, usec: parts.fourth, utc: true}
      end
    end

    def extract_hour_minute_second_parts(str)
      # match array returns ["12:35:20 AM", "12", "35", "20", "AM"]
      parts = str.match(HOUR_MINUTE_SECOND_REGEX).to_a.pop(4)
      if parts.present?
        {hour: make_hour(parts.first.to_i, parts.fourth), min: parts.second.to_i, sec: parts.third.to_i}
      end
    end

    def extract_hour_minute_parts(str)
      # match array returns ["12:35 AM", "12", "35", "AM"]
      parts = str.match(HOUR_MINUTE_REGEX).to_a.pop(3)
      if parts.present?
        {hour: make_hour(parts.first.to_i, parts.third), min: parts.second.to_i}
      end
    end

    # Convert 2 digit years into 4
    def make_year(year)
      return year if year.to_s.size > 2

      start_year = time_zone.now.year - Flexitime.configuration.ambiguous_year_future_bias
      century = (start_year / 100) * 100
      full_year = century + year
      full_year < start_year ? full_year + 100 : full_year
    end

    # Convert hour depending on presence of am/pm
    def make_hour(hour, meridiem)
      meridiem.to_s.downcase == "pm" ? hour + 12 : hour
    end

    # Validate the day & month parts as Time.zone#local accepts some invalid values
    # such as Time.zone.local(2021,2,30) returning "2021-03-02"
    def valid_date_parts?(parts)
      parts[:month] >= 1 && parts[:month] <= 12 && parts[:day] >= 1 &&
        parts[:day] <= Time.days_in_month(parts[:month], parts[:year])
    end

    # Create a TimeWithZone object object using only those parts required for the configuration precision
    def create_time_from_parts(parts, precision: nil)
      time = time_zone.local(*local_args_for_precision(parts, precision: precision))
      parts[:utc] ? (time + time.utc_offset) : time
    rescue
    end

    # Returns the date/time parts required for the configuration precision
    def local_args_for_precision(parts, precision: nil)
      keys = [:year, :month, :day, :hour, :min, :sec, :usec]
      precision ||= Flexitime.configuration.precision
      index = keys.index(precision)
      keys[0..index].map { |key| parts[key] }
    end

    # Parse the string using Time.zone and set the configuration precision
    def time_zone_parse(str, precision: nil)
      time = time_zone.parse(str)
      set_precision(time, precision: precision)
    rescue
    end

    # Set the precision, first checking if this is necessary
    # to avoid the overhead of calling the change method
    def set_precision(time, precision: nil)
      precision ||= Flexitime.configuration.precision
      index = PRECISIONS.index(precision)
      dismiss_part = PRECISIONS[index + 1]
      excess = PRECISIONS[(index + 1)..-1].sum { |key| time.send(key) }
      excess > 0 ? time.change(dismiss_part => 0) : time
    end
  end
end


lib/flexitime/configuration.rb
# frozen_string_literal: true

module Flexitime
  PRECISIONS = [:day, :hour, :min, :sec, :usec].freeze
  DATE_PARTS = [:day, :month].freeze

  class Configuration
    # = Configuration options
    #
    # == first_date_part
    # The first part of the date within the string, either :day or :month
    # This can be set manually otherwise when using the rails-I18n gem if the first element
    # of the "date.order" translation is :day or :month that will be used
    # otherwise the default of :day will be used
    #
    #   Flexitime.first_date_part = :day
    #   Flexitime.parse("01/06/2021") # => 2021-06-01 00:00:00 +0100
    #
    #   Flexitime.first_date_part = :month
    #   Flexitime.parse("01/06/2021") # => 2021-01-06 00:00:00 +0100
    #
    # Regardless of the option value, the gem will always attempt to parse
    # a string starting with a 4 digit year, deciphering the format as year/month/day
    #
    #   Flexitime.parse("2021-11-12 08:15") # => 2021-11-12 08:15:00 +0000
    #
    # == precision
    # The desired precision for the returned Time object, defaulting to minute (:min)
    #
    #   Flexitime.parse("01/06/2021 18:30:45.036711") # => 2021-06-01 18:30:00 +0100
    #   Flexitime.precision = :sec
    #   Flexitime.parse("01/06/2021 18:30:45.036711") # => 2021-06-01 18:30:45 +0100
    #   Flexitime.precision = :usec
    #   Flexitime.parse("01/06/2021 18:30:45.036711") # => 2021-06-01 18:30:45.036711 +0100
    #
    # == ambiguous_year_future_bias
    # The option used to determine the century when parsing a string containing a 2 digit year
    # with the default value of 50 and the current year of 2020
    # the year is set from within a range of >= 1970 and <= 2069
    # whereas with a value of 20 and the current year of 2020
    # the year is set from within a range of >= 2000 and <= 2099
    #
    #   Flexitime.parse("01/08/00").year # => 2000
    #   Flexitime.parse("01/08/69").year # => 2069
    #   Flexitime.parse("01/08/70").year # => 1970
    #   Flexitime.parse("01/08/99").year # => 1999
    #
    attr_reader :precision
    attr_accessor :ambiguous_year_future_bias

    def initialize
      @first_date_part = nil
      @precision = :min
      @ambiguous_year_future_bias = 50
    end

    def first_date_part=(first_date_part)
      raise ArgumentError.new("Invalid first date part") unless first_date_part_valid?(first_date_part)
      @first_date_part = first_date_part.to_sym
    end

    def first_date_part
      first_date_part = @first_date_part || i18n_first_date_part
      first_date_part_valid?(first_date_part) ? first_date_part : DATE_PARTS.first
    end

    def precision=(precision)
      raise ArgumentError.new("Invalid precision") unless precision_valid?(precision)
      @precision = precision.to_sym
    end

    private

    def i18n_first_date_part
      Array(I18n.t("date.order")).map(&:to_sym).first if defined?(I18n)
    end

    def first_date_part_valid?(first_date_part)
      first_date_part.present? && DATE_PARTS.include?(first_date_part.to_s.to_sym)
    end

    def precision_valid?(precision)
      precision.present? && PRECISIONS.include?(precision.to_s.to_sym)
    end
  end
end
